tools/gfx/debug-layer.cpp

```diff
@@ -669,6 +669,42 @@ Result DebugDevice::createMutableRootShaderObject(
     return result;
 }
 
+Result DebugDevice::createShaderObjectFromTypeLayout(
+    slang::TypeLayoutReflection* typeLayout, IShaderObject** outShaderObject)
+{
+    SLANG_GFX_API_FUNC;
+
+    RefPtr<DebugShaderObject> outObject = new DebugShaderObject();
+    auto result = baseObject->createShaderObjectFromTypeLayout(typeLayout, outObject->baseObject.writeRef());
+    auto type = typeLayout->getType();
+    auto typeName = type->getName();
+    outObject->m_typeName = typeName;
+    outObject->m_device = this;
+    outObject->m_slangType = type;
+    if (SLANG_FAILED(result))
+        return result;
+    returnComPtr(outShaderObject, outObject);
+    return result;
+}
+
+Result DebugDevice::createMutableShaderObjectFromTypeLayout(
+    slang::TypeLayoutReflection* typeLayout, IShaderObject** outShaderObject)
+{
```

Comment:
It seems like we will eventually want to have most of the exposed `create*ShaderObject*()` operations bottleneck through a single bottleneck routine to make things like the `DebugDevice` less onerous to implement.

I suppose that could either be done by having some of those operations just be helper/wrapper operations that sit above the binary interface of the COM objects, or by using even more subclassing in the implementations such that even `DebugDevice` can derive some of these implementations such that they bottleneck into one another.